### ISSUE TITLE
Discussion: Queries on array-indexed objects

### ISSUE THREAD

**Issue by CoDEmanX** [arangodb/arangodb#1502]
This is a follow-up on a brief discussion I had with @mchacki at Distributed Matters conference about particular queries (objects nested in arrays) in conjunction with the upcoming array indexes.

Example document:

``` js
{
    "_key": 75,
    "objectName": "GS Kunst 3138",
    "title": "Vuë d´un combat naval",
    "categories": [
        "Grafik",
        "Militaria",
        "Schifffahrt"
    ],
    "administrationName": "Abteilung Tiefdrucke",
    "production": {
        "dateEnd": 1790,
        "dateStart": 1766,
        "period": "2. Hälfte 18. Jh."
    },
    "creator": {
        "name": "Probst, Georg Balthasar"
    },
    "dimensions": [
        {
            "type": "Höhe",
            "part": "Bild",
            "unit": "cm",
            "value": 26.6
        },
        {
            "type": "Breite",
            "part": "Bild",
            "unit": "cm",
            "value": 39.6
        },
        {
            "type": "Höhe",
            "part": "Blatt",
            "unit": "cm",
            "value": 29.6
        },
        {
            "type": "Breite",
            "part": "Blatt",
            "unit": "cm",
            "value": 41
        }
    ],
    "reproduction": {
        "referenceNumber": "33_00009021.jpg"
    },
    "input": {
        "user": "j.doe",
        "date": "2002-10-25"
    }
}
```

Array index: at least on `dimensions[*].type`

Desired query: find documents whose height (Höhe) is greater than 40 cm, pseudo query:

```
dimensions[*].type == "Höhe" AND dimensions.unit[*] == "cm" AND dimensions[*].value > 40
```

The problem: the following document part matches the `value`-criterion

``` js
        {
            "type": "Breite"
            "part": "Blatt",
            "unit": "cm",
            "value": 41
        }
```

because `value > 40` is true, but it's not of type "Höhe". Another dimension element contains `type == "Höhe"`, hence the document matches. The actual intention of the query is to match all criteria within the **same array element** / object however, and the example document does not satisfy all 3 conditions in a single dimensions object.

A new keyword could be introduced to tell the query executor to find `dimensions.type == "Höhe"` using the array index, and then perform additional checks on the matching array elements only:

```
dimensions[*].type == "Höhe" WHEN dimensions[*].unit == "cm" WHEN dimensions[*].value > 40
```

Or another crazy idea:

```
FOR doc IN collection
    WITH doc.dimensions AS d
        FILTER d.type == "Höhe" AND d.unit == "cm" AND d.value > 40
        RETURN doc
    //WITH_END?
```

AQL doesn't actually support a statement like `doc.dimensions[*].type == "Höhe"` however, and `WITH` is ugly (could just use a FOR loop). The following **does** work in current AQL:

``` js
FOR doc IN test
    FOR d IN doc.dimensions
        FILTER d.type == "Höhe" AND d.unit == "cm" AND d.value > 40
        COLLECT a = doc._key INTO b
        RETURN b[0].doc
```

From 2.7 on, the COLLECT/RETURN hack can be replaced by a simple `RETURN DISTINCT doc` (DISTINCT prevents a document to be returned multiple times caused by the dimensions FOR loop).

My question: can the array index on `.type` be utilized in such a query, with a for-loop over the dimensions array?

An alternative way to formulate the query I want (works in 2.7 devel):

``` js
FOR doc IN collection
    FILTER LENGTH(doc.dimensions[* FILTER CURRENT.type == "Höhe" AND CURRENT.value > 40])
    RETURN doc
```

An inline-filter is used, which will only return array elements of the _dimensions_ attribute if all conditions are true. If there's at least one matching dimension, the FILTER criterion will evaluate to true (LENGTH != 0) and return the document - without the need for a loop, `COLLECT` or `RETURN DISTINCT`!

It seems way more natural and flexible then the earlier proposed `WHEN` operator. Question is again, if array indexes can/will be used to make this query efficient.

Furthermore, I wonder if above query could be made faster by explicitly returning `1` inside the inline filter to speed it up (we aren't interested in what CURRENT actually contains) - or is there already an optimizer rule to avoid unnecessary copying of data to the outer FILTER?

---

As a side note: `RETURN DISTINCT` doesn't support the expansion operator it seems, but it would actually be quite handy.

``` js
FOR doc IN test
    RETURN DISTINCT doc.dimensions[*].type

  [
    "Höhe",
    "Breite",
    "Höhe",
    "Breite"
  ]
```

vs.

``` js
FOR doc IN test
    FOR dimension IN doc.dimensions
        COLLECT type = dimension.type
        RETURN  type

[
  "Breite",
  "Höhe"
]
```

**Comment by jsteemann** [comment]
The `RETURN DISTINCT` in the last example is not applied on the value `doc.dimensions[*].type` (so it may be an array with duplicate values), but on the results produced by the `FOR` loop. So the for loop will not return the same two values.
For example, if there are 4 documents with these values in `doc.dimensions[*].type`: `[ 1, 2, 3 ]`, `[ 1, 1, 2, 3 ]`, `[ 3, 2, 1 ]` and `[ 1, 2, 3 ]`, then the `FOR` loop will produce the following result: `[ 1, 2, 3 ]`, `[ 1, 1, 2, 3 ]` and `[ 3, 2, 1 ]`. One element (the last) will get removed because it's identical to the first. The second element will not get modified, even though it contains duplicate elements. This is because of the `DISTINCT` being applied on the `FOR` loop level.

**Comment by CoDEmanX** [comment]
You're right, I imagined a single document only. To apply DISTINCT on the values, something like this would be required:

    RETURN DISTINCT[*]  doc.dimensions[*].type

**Comment by dothebart** [comment]
http://stackoverflow.com/questions/37029697/query-nested-arrays-in-arangodb/37115219?noredirect=1#comment61989851_37115219 has a similiar topic.

**Comment by gghdev** [comment]
Hi there. This is an old thread but it exactly matches what I'm trying to. The only variation I have is that I'm trying to do this in ArangoSearch. The following example works. It finds a document that has a subarray containing _both_ a given tag_id (123) _and_ value (456). The additional FILTER -- albeit the fastest solution I've found -- doesn't seem optimal:

`FOR doc IN page_view
SEARCH doc.tags.tag_id == 123 AND doc.tags.values == 456
FILTER LENGTH(doc.tags[* FILTER CURRENT.tag_id == 123 AND CURRENT.values ANY == 456])
RETURN doc`

Is there, or could there be, a better way to SEARCH by two attributes in the same subarray, without resorting to FILTERing? Thanks.

**Comment by Simran-B** [comment]
CC @gnusi @Dronplane

**Comment by Simran-B** [comment]
Related: #10583 #10847

**Comment by gnusi** [comment]
@gghdev we plan to add a functionality to allow efficient search within sub-objects, but unfortunately there is not ETA atm.

**Comment by Simran-B** [comment]
Meanwhile, there is the so-called question mark operator and support for it in ArangoSearch:
https://docs.arangodb.com/stable/aql/operators/#question-mark-operator

From 3.10.0 onward:
https://docs.arangodb.com/stable/release-notes/version-3.10/whats-new-in-3-10/#question-mark-operator
